xcashu/server/middleware.py

```diff
@@ -0,0 +1,130 @@
+import json
+import base64
+
+from fastapi.responses import JSONResponse
+from starlette.middleware.base import BaseHTTPMiddleware
+
+from cashu.core.base import TokenV2
+from cashu.core.settings import LIGHTNING
+from cashu.core.helpers import sum_proofs
+
+from xcashu.server.ledger import ledger
+
+
+class EcashHeaderMiddleware(BaseHTTPMiddleware):
+    """
+    Middleware that checks the HTTP headers for ecash
+    """
+
+    async def dispatch(self, request, call_next):
+        if is_payment_required(request.url.path):
+            # check whether valid ecash was provided in header
+            token = get_payment_token(request)
+            required_amount = get_required_amount(request.url.path)
+            if not token:
+                return await payment_information(get_invoice_amount(request))
+
+            error = await handle_payment(token, required_amount)
+            if error:
+                return error
+
+        return await call_next(request)
+
+
+def is_payment_required(path: str) -> bool:
+    """
+    Check whether the path requires ecash
+    """
+    return get_required_amount(path) > 0
+
+
+def get_required_amount(path: str) -> int:
+    """
+    Get the required amount for the path
+    """
+    # TODO: get paid paths and required amount from config
+    # all requests to /cashu/* are not checked for ecash
+    if path.startswith("/paid/"):
+        return 1
+    return 0
+
+
+def get_payment_token(request) -> str:
+    """
+    Get the payment token from the request
+    """
+    # TODO: get header from config?
+    # TODO: a more "generic" way could be to introduce a "Payment" header
+    # and then expect a value "Cashu <payment token>"
+    # Example:
+    # ```
+    # HTTP POST /paid/endpoint
+    # Payment: Cashu <payment token>
+    # ```
+    # Similar to how the `Authorization` header works in HTTP
+    # (Example: `Authorization: Bearer <token>`)
```

Comment:
Some ideas:
- get "header" name from a config and fallback to `X-Cashu`
- or a more "generic" one that still indicates that a given token is meant to be used with a Cashu ecash payment system?